ShareFileJavaSDK/src/com/citrix/sharefile/api/SFApiQueryExecutor.java

```diff
@@ -209,137 +210,157 @@ private T executeQueryWithReAuthentication() throws SFServerException,
                     }
-                    //break;
+                    return ret;
+                }
 
-                    case HttpsURLConnection.HTTP_NO_CONTENT:
-                    {
-                        return null;
-                    }
-                    //break;
+                case HttpsURLConnection.HTTP_NO_CONTENT:
+                {
+                    return null;
+                }
 
-                    case HttpsURLConnection.HTTP_UNAUTHORIZED:
-                    {
-                        Logger.d(TAG, "RESPONSE = AUTH ERROR");
+                case HttpsURLConnection.HTTP_UNAUTHORIZED:
+                {
+                    Logger.d(TAG, "RESPONSE = AUTH ERROR");
 
-                        callWipeCredentialsCallback();
+                    callWipeCredentialsCallback();
 
-                        SFFormsAuthenticationCookies formsAuthResponseCookies = SFHttpsCaller.getFormsAuthResponseCookies(url, connection, mCookieManager);
-                        if(formsAuthResponseCookies != null) {
-                            throw new SFNotAuthorizedException(SFKeywords.UN_AUTHORIZED, formsAuthResponseCookies, mReAuthContext);
-                        }
-                        return executeQueryWithReAuthentication();
+                    SFFormsAuthenticationCookies formsAuthResponseCookies = SFHttpsCaller.getFormsAuthResponseCookies(url, connection, mCookieManager);
+                    if(formsAuthResponseCookies != null) {
+                        throw new SFNotAuthorizedException(SFKeywords.UN_AUTHORIZED, formsAuthResponseCookies, mReAuthContext);
                     }
-                    //break;
-
-                    /*
-                       The ShareFile server doesn't treat 404 correctly. Eg: Try creating a
-                       duplicate folder on ShareFile. The resulting http code
-                       is ShareFile(404) vs Connectors(409).
-                       So attempt to parse the server response String for accurate error message.
-                    */
-                    case HttpsURLConnection.HTTP_NOT_FOUND:
-                    {
-                        responseString = SFHttpsCaller.readErrorResponse(connection);
-                        Logger.v(TAG, responseString);
-                        SFV3ErrorParser sfV3error = new SFV3ErrorParser(httpErrorCode, responseString, null);
-                        throw new SFNotFoundException(sfV3error.errorDisplayString());
+                    // Writelock to prevent use of token while it's being updated, note state before
+                    String tokenBefore = mSFApiClient.getOAuthToken().getAccessToken();
+                    mSFApiClient.readWriteLock.writeLock().lock();
+                    if (! tokenBefore.equals(mSFApiClient.getOAuthToken().getAccessToken())) {
+                        // If another thread already refreshed while this thread was waiting
+                        return executeBlockingQuery();
```

Comment:
I think we need to release the write lock before doing this?

Reply:
Good point. It should really just be around the `reauthenticate` call. I'll move it.

Reply:
Well refetching the auth token needs to happen inside the write lock, right? Unless we're ok with several threads reauthenticating.

Reply:
Yeah it can still let some through. How about:
```java
// Writelock to prevent use of token while it's being updated
mSFApiClient.readWriteLock.writeLock().lock();
try {
    if (tokenBefore.equals(mSFApiClient.getOAuthToken().getAccessToken())) {
        // If no other thread has already updated token
        // Attempt to reauthenticate, if no exception, retry query
        reauthenticate();
    }
} finally {
    mSFApiClient.readWriteLock.writeLock().unlock();
}
return executeBlockingQuery();
```
If the reauth doesn't work it will throw an exception and should not do the recursive call

Reply:
Yeah looks good to me!